### PR TITLE
feat: support operation validation via `check` command

### LINKED ISSUE #416: Pass `operations` to `check` command
**Problem**
As a user I wish to validate the generated graphql schema with a set of graphQL queries.

**Solution 1**
Since #438 we now support loading multiple graphQL files. Which means that we can pass the queries as a part of the configurations more easily. We should just validate the queries that are in the graphQL files without and special check parameter.

The command would look a lot simpler now:
```bash
tc check ./jsonplaceholder.graphql query1.graphql query2.graphql
```
All the above options should be allowed to test the graphQL schema against the provided queries and mutations. 

**Solution 2**
We will extend the check command on the CLI to support checking of the schema against provided queries and mutations:

```bash
# using operation parameter
tc check ./jsonplaceholder.graphql --operation query1.graphql --operation mutation1.graphql

# using comma as a delimiter
tc check ./jsonplaceholder.graphql --operation file1.graphql,file2.graphql

# using space as a delimeter
tc check ./jsonplaceholder.graphql --operation file1.graphql file2.graphql
```
All the above options should be allowed to test the graphQL schema against the provided queries and mutations.

**Technical Requirements**
- Use `ValidationError` to capture and display all the errors.
- Errors should be composed with errors `from_config`.

### PR DESCRIPTION
**Pull request by tobihans**
**Summary:**  
This PR add ability to check that the schema is able to execute some operations. It follows the second approach outlined in the [issue](https://github.com/tailcallhq/tailcall/issues/416#issue-1925530012).

It's worth noting that the validation depends on the `queryValidation` parameter of the `@server` directive, as the query is executed against the schema (with resolvers that don't actually do anything).

Given a the sample schema `jsonplaceholder.graphql` and two operations `posts.graphql` and `empty.graphql`, we get the output below.
```bash
# tc check jsonplaceholder.graphql --operations posts.graphql empty.graphql

N + 1: 1

Error: Validation Error
• Unknown field "userOd" on type "Post". Did you mean "userId", "user"? [examples/operations/posts.graphql, line 5 column 5]
• Unknown field "userAd" on type "Post". Did you mean "userId", "user"? [examples/operations/posts.graphql, line 6 column 5]
•  --> 1:1
  |
1 | 
  | ^---
  |
  = expected executable_definition [examples/operations/empty.graphql, line 1 column 1]


```

- `jsonplaceholder.graphql`
```graphql
schema
  @server(port: 8000, graphiql: true, hostname: "0.0.0.0", queryValidation: true)
  @upstream(baseURL: "http://jsonplaceholder.typicode.com", httpCache: true) {
  query: Query
}

type Query {
  posts: [Post] @http(path: "/posts")
}

type User {
  id: Int!
  name: String!
  username: String!
  email: String!
  phone: String
  website: String
}

type Post {
  id: Int!
  userId: Int!
  title: String!
  body: String!
  user: User @http(path: "/users/{{value.userId}}")
}
```
- `posts.graphql`
```graphql
query {
  posts {
    id
    body
    userOd
    userAd
  }
}
```
- `empty.graphql`: well, it's empty

**Issue Reference(s):**  
Fixes #416 

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [x] I have updated the documentation accordingly (if applicable).
- [x] I have performed a self-review of my own code.

**Fun fact:**

I mistakenly deleted the branch, which closed the first [PR](https://github.com/tailcallhq/tailcall/pull/692) I opened.

/claim #416

### CODE REVIEW

File: src/cli/command.rs [hunk 1]
```diff
@@ -42,6 +42,10 @@ pub enum Command {
     /// Display schema
     #[arg(short, long)]
     schema: bool,
+
+    /// Operations to check
+    #[arg(short, long, num_args = 1..)]
+    operations: Vec<String>,
```
Comment (b4s36t4) [inline src/cli/command.rs:48]: Add [value_delimiter](https://docs.rs/clap/latest/clap/struct.Arg.html#method.value_delimiter) this to the arg so that even `,` comma separated values get captured.

**Review by b4s36t4 — CHANGES_REQUESTED** [review]

File: examples/jsonplaceholder.graphql [hunk 2]
```diff
@@ -1,5 +1,5 @@
 schema
-  @server(port: 8000, graphiql: true, hostname: "0.0.0.0")
+  @server(port: 8000, graphiql: true, hostname: "0.0.0.0", queryValidation: true)
```
Comment (ologbonowiwi) [inline examples/jsonplaceholder.graphql:2]:
```suggestion
  @server(port: 8000, graphiql: true, hostname: "0.0.0.0")
```

File: src/blueprint/into_schema.rs [hunk 3]
```diff
@@ -39,26 +39,29 @@ fn to_type(def: &Definition) -> dynamic::Type {
...
-                  a => FieldValue::from(a),
-                };
-                Ok(Some(p))
-              }
-            }
-          })
+            })
+          } else {
+            FieldFuture::new(async { Ok(None::<FieldValue>) })
+          }
```
Comment (ologbonowiwi) [inline src/blueprint/into_schema.rs:64]: Is this an actual change, or was it committed by mistake?

File: src/cli/tc.rs [hunk 4]
```diff
@@ -104,11 +108,37 @@ pub async fn init(file_path: &str) -> Result<()> {
...
+
+fn validate_operations(blueprint: &Blueprint, operations: Vec<String>) -> Result<()> {
+  tokio::runtime::Runtime::new()?.block_on(async {
+    let schema = blueprint.to_schema();
+    let mut execution = vec![];
+
+    for op in operations.iter() {
+      if let Ok(operation) = tokio::fs::read_to_string(op).await {
+        // Fake execution to get errors
+        let Response { errors, .. } = schema.execute(&operation).await;
```
Comment (ologbonowiwi) [inline src/cli/tc.rs:127]: We don't have a way to validate errors without executing the schema?

File: src/cli/tc.rs [hunk 5]
```diff
@@ -104,11 +108,37 @@ pub async fn init(file_path: &str) -> Result<()> {
...
   Fmt::display(Fmt::table(seq));
 }
+
+fn validate_operations(blueprint: &Blueprint, operations: Vec<String>) -> Result<()> {
+  tokio::runtime::Runtime::new()?.block_on(async {
+    let schema = blueprint.to_schema();
+    let mut execution = vec![];
+
+    for op in operations.iter() {
+      if let Ok(operation) = tokio::fs::read_to_string(op).await {
```
Comment (ologbonowiwi) [inline src/cli/tc.rs:125]: Instead of adding an `if else` block here to check if it is ok, can you use `unwrap_or_else` instead? it'd make it more readable

File: src/cli/tc.rs [hunk 6]
```diff
@@ -104,11 +108,37 @@ pub async fn init(file_path: &str) -> Result<()> {
...
+        // Fake execution to get errors
+        let Response { errors, .. } = schema.execute(&operation).await;
+        execution.push((op, errors));
+      } else {
+        return Err(anyhow!("Cannot read operation {}", op));
+      }
+    }
```
Comment (ologbonowiwi) [inline src/cli/tc.rs:132]: Instead of creating a mut variable and pushing to it, we could use `.map` and assign it directly

e.g.:
```suggestion
    let mut execution = vec![];
    let execution = operations.iter()
        .map(/* the operation processing */)
```

Comment (ologbonowiwi) [reply src/cli/tc.rs:132]: also, once we do this, we can map into `format_operation` directly, and on `Fmt::display`, we'll need only to `join(\n\n)` (and possibly `.collect::<Vec<String>>()`)

File: src/cli/command.rs (hunk 1 above)
Comment (tobihans) [reply src/cli/command.rs]: Thank you.

File: src/blueprint/into_schema.rs (hunk 3 above)
Comment (tobihans) [reply src/blueprint/into_schema.rs:64]: That's a change. Because I want to avoid actually running the execution logic when checking the operation. Otherwise, checking will run real requests.****

File: src/cli/tc.rs (hunk 4 above)
Comment (tobihans) [reply src/cli/tc.rs:127]: Well, I tried at first. But that would just be too much work and I believe it's out of the scope of the project. `async_graphql` is already parsing well, so creating an `Operation` struct to handle all that, would be a lot of duplicated efforts. But, because the resolvers are fake when checking, the schema execution doesn't return any meaning result, yet it still performs all the required checks.

Comment (tobihans) [reply src/cli/tc.rs:127]: I also thought about parsing the operation then validating the fields manually, but then, it's still repetitive.

File: src/cli/tc.rs (hunk 5 above)
Comment (tobihans) [reply src/cli/tc.rs:125]: By design, I exit if we fail to read the operation file. There is nothing to provide if we cannot read the op, I think.

File: src/cli/tc.rs (hunk 6 above)
Comment (tobihans) [reply src/cli/tc.rs:132]: But then, it will require to handle failures in reading in another way. I'll look into it. Thanks

File: src/cli/tc.rs [hunk 7]
```diff
@@ -104,11 +108,37 @@ pub async fn init(file_path: &str) -> Result<()> {
 pub fn display_schema(blueprint: &Blueprint) {
   Fmt::display(Fmt::heading(&"GraphQL Schema:\n".to_string()));
   let sdl = blueprint.to_schema();
-  Fmt::display(print_schema::print_schema(sdl));
+  Fmt::display(format!("{}\n", print_schema::print_schema(sdl)));
```
Comment (ologbonowiwi) [inline src/cli/tc.rs:111]: these changes are out of the scope of #416

```suggestion
  Fmt::display(print_schema::print_schema(sdl));
```

Comment (ologbonowiwi) [reply src/cli/tc.rs:111]: I got that this meant to break the line, but we need to do so not here but where we'll print the other message

File: src/cli/tc.rs [hunk 8]
```diff
@@ -104,11 +108,37 @@ pub async fn init(file_path: &str) -> Result<()> {
...
 }
 
 fn display_config(config: &Config, n_plus_one_queries: bool) {
-  Fmt::display(Fmt::success(&"No errors found".to_string()));
```
Comment (ologbonowiwi) [inline src/cli/tc.rs:111]: We need to restore this; the message should be changed but not deleted

```suggestion
  Fmt::display(Fmt::success(&"No errors found on config".to_string()));
```

File: src/cli/tc.rs (hunk 6 above)
Comment (ologbonowiwi) [reply src/cli/tc.rs:132]: If you handle the failures as mentioned [here](https://github.com/tailcallhq/tailcall/pull/706/files#r1411120957) (with `unwrap_or_else`), you can do it with only a few tweaks.

**Review by ologbonowiwi — CHANGES_REQUESTED** [review]
We need to find a way to write automated tests to this feature

File: src/cli/tc.rs (hunk 5 above)
Comment (meskill) [reply src/cli/tc.rs:125]: the example above is not right, `unwrap_or_else` is used to provide defaults if we have an error and will not end the function with the result. 

The current code is valid if we want to fail fast on errors. To simplify it we can use something like this:

```
let value = func().map_err(...)?;
```

File: src/blueprint/into_schema.rs (hunk 3 above)
Comment (meskill) [reply src/blueprint/into_schema.rs:64]: agree with @ologbonowiwi  - it's a good concern to not execute request for validation, but check like this is too implicit and error prone, so we would like to have some kind of explicit switch or function to have that custom logic

File: src/cli/tc.rs [hunk 9]
```diff
@@ -104,11 +108,35 @@ pub async fn init(file_path: &str) -> Result<()> {
...
   let seq = vec![Fmt::n_plus_one_data(n_plus_one_queries, config)];
   Fmt::display(Fmt::table(seq));
 }
+
+fn validate_operations(blueprint: &Blueprint, operations: Vec<String>) -> Result<()> {
```
Comment (tusharmath) [inline src/cli/tc.rs:119]: Can be moved to `blueprint`. Instead of result it should return a `Valid<(), String>`

File: src/cli/fmt.rs [hunk 10]
```diff
@@ -40,6 +44,23 @@ impl Fmt {
...
+    let heading = Fmt::heading(&format!("Operation {op}: {meta}"));
+
+    let report = errors
+      .iter()
+      .enumerate()
+      .map(|(count, error)| Fmt::error(&format!("{}. {}", count + 1, error.message)))
+      .collect::<Vec<String>>()
+      .join("\n");
+
+    format!("{heading}\n{report}")
```
Comment (tusharmath) [inline src/cli/fmt.rs:62]: If you return a `Valid` it can be easily handled by the CLI internally because it's built on a ValidationError which is converted to a `CLIError` which is CLI friendly.

File: src/cli/tc.rs [hunk 11]
```diff
@@ -104,11 +108,35 @@ pub async fn init(file_path: &str) -> Result<()> {
...
+  tokio::runtime::Runtime::new()?.block_on(async {
+    let schema = blueprint.to_validation_schema();
+    let mut execution = vec![];
+
+    for op in operations.iter() {
+      let operation = tokio::fs::read_to_string(op)
+        .await
+        .map_err(|_| anyhow!("Cannot read operation {}", op))?;
+        let Response { errors, .. } = schema.execute(&operation).await;
+        execution.push((op, errors));
```
[the remark this answers is not in the document]
Comment (tobihans) [reply src/cli/tc.rs:129]: Yeah, it's not real execution. I create a fake schema with `blueprint.to_validation_schema`. It doesn't make real http calls. Thanks

File: src/cli/tc.rs [hunk 12]
```diff
@@ -104,11 +108,35 @@ pub async fn init(file_path: &str) -> Result<()> {
...
+    }
```
Comment (ologbonowiwi) [inline src/cli/tc.rs:130]: Now that we're handling errors with `map_err`, we can change this loop to a map. The code may require some tweaks to work 

```suggestion
    let execution = operations
      .iter()
      .map(async |op| {
        let operation = tokio::fs::read_to_string(op)
          .await
          .map_err(|_| anyhow!("Cannot read operation {}", op))? //This can likely be a CLIError
          
        let Response { errors, .. } = schema.execute(&operation).await;
        (op, errors)
      });
```

**Review by ologbonowiwi — CHANGES_REQUESTED** [review]

File: src/blueprint/blueprint.rs [hunk 13]
```diff
@@ -210,4 +214,29 @@ impl Blueprint {
...
+      match tokio::fs::read_to_string(op).await {
+        Ok(operation) => {
+          let Response { errors, .. } = schema.execute(&operation).await;
+          execution.push((op, errors.iter().map(|e| e.message.clone()).collect::<Vec<String>>()));
+        }
+        Err(_) => execution.push((op, vec!["Cannot read operation".to_string()])),
+      }
+    }
+
+    Valid::from_iter(execution.iter(), |(op, errors)| Fmt::format_operation(op, errors)).map(|_| ())
```
Comment (tobihans) [inline src/blueprint/blueprint.rs:240]: Please, I'm very unfamiliar with the `Valid` struct and still trying to understand it. I don't know if someone can explain me why this results in duplicated errors in the console like this:
```bash
Error: Validation Error
• Operation examples/operations/posts.graphql: 2 errors found
1. Unknown field "userOd" on type "Post". Did you mean "userId", "user"?
2. Unknown field "userAd" on type "Post". Did you mean "userId", "user"? []
• Operation examples/operations/empty.graphql: 1 errors found
1. Cannot read operation []
Validation Error
• Operation examples/operations/posts.graphql: 2 errors found
1. Unknown field "userOd" on type "Post". Did you mean "userId", "user"?
2. Unknown field "userAd" on type "Post". Did you mean "userId", "user"? []
• Operation examples/operations/empty.graphql: 1 errors found
1. Cannot read operation []
```
Thanks

Comment (ologbonowiwi) [reply src/blueprint/blueprint.rs:240]: You're using `Valid` here and then using `Valid::fail` on `format_operation`. Maybe that's the issue?

File: src/blueprint/into_schema.rs [hunk 14]
```diff
@@ -133,6 +133,42 @@ fn create(blueprint: &Blueprint) -> SchemaBuilder {
   schema
 }
 
+pub(crate) fn create_dumb_schema(blueprint: &Blueprint) -> SchemaBuilder {
```
[the remark this answers is not in the document]
Comment (tobihans) [reply src/blueprint/into_schema.rs:136]: It required to alter the behavior of the `create` function, which might lead to bugs, and updating the blueprint would still not prevent the real execution of the requests, which is why I added this method. But it reuses most of the previous code. It only makes _null_ the resolver.

**Comment by tobihans** [comment]
Thanks for the branch update @tusharmath. I'm currently really sick. I hope to feel better soon and complete this. It remains to write automated tests like in `graphql_spec.rs`

File: src/blueprint/into_schema.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (meskill) [reply src/blueprint/into_schema.rs:136]: I totally get why you've added the method. 

But to express my concerns in more details: there is no code reuse here because you've actually created new code instead of using previous one. We can only hope that code acts the same way the original code works. For now it could be easy to see it's ok, but such code tends to diverge with time and it needs to be supported. Moreover, see the codecov comments in your pr? It shows this function is not covered by tests, so we probably need to create additional set of tests to make sure `create_dumb_schema` works as expected that should be same as tests for normal workflow

**Comment by meskill** [comment]
@tobihans Hey, could you please update us on the status of the pr?

**Comment by tobihans** [comment]
> @tobihans Hey, could you please update us on the status of the pr?

I'll ask for a review in one day or two. Thanks.

File: src/cli/tc.rs (hunk 7 above)
Comment (ologbonowiwi) [reply src/cli/tc.rs:156]: this review is still valid @tobihans

File: src/cli/tc.rs (hunk 8 above)
Comment (ologbonowiwi) [reply src/cli/tc.rs:147]: this review is still valid @tobihans

File: npm/gen-root.ts [hunk 15]
```diff
@@ -73,7 +73,7 @@ async function genServerPackage(buildDefinitions: string[]) {
 
   const postInstallScriptContent = `const version = "${packageVersion}";\n${postInstallScript}`
   const preInstallScriptContent = `const optionalDependencies = ${JSON.stringify(
-    optionalDependencies,
```
Comment (ologbonowiwi) [inline npm/gen-root.ts:76]: we shouldn't change this as does not relate with the issue being fixed

File: npm/gen-root.ts [hunk 16]
```diff
@@ -73,7 +73,7 @@ async function genServerPackage(buildDefinitions: string[]) {
...
+    optionalDependencies
```
Comment (ologbonowiwi) [inline npm/gen-root.ts:76]: we shouldn't change it as it does not relate with the issue being solved
```suggestion
    optionalDependencies,
```

File: src/blueprint/blueprint.rs [hunk 17]
```diff
@@ -219,4 +221,53 @@ impl Blueprint {
...
+        Ok(operation) => {
+          let causes = Blueprint::validate_operation(&schema, &operation).await;
+          execution.push((op.to_string(), causes));
+        }
+        Err(_) => execution.push((
+          op.to_string(),
+          vec![Cause::new(format!("Cannot read file operation file {}", op))],
+        )),
+      }
+    }
```
Comment (ologbonowiwi) [inline src/blueprint/blueprint.rs:266]: we should use `operations.iter().map()` rather than using for and `.push` values into execution.

File: src/cli/mod.rs [hunk 18]
```diff
@@ -4,4 +4,5 @@ mod fmt;
 mod tc;
 
 pub use error::CLIError;
+pub use fmt::Fmt;
```
Comment (ologbonowiwi) [inline src/cli/mod.rs:7]: we do use it outside `cli` module? if not, we don't need the `pub` identifier (we can use `pub(crate)` instead)

File: src/cli/tc.rs (hunk 7 above)
Comment (ologbonowiwi) [reply src/cli/tc.rs:156]: this review is still valid @tobihans

File: src/cli/tc.rs (hunk 8 above)
Comment (ologbonowiwi) [reply src/cli/tc.rs:147]: this review is still valid @tobihans. IMO, we shouldn't delete this line.

**Review by ologbonowiwi — CHANGES_REQUESTED** [review]

File: src/blueprint/blueprint.rs [hunk 19]
```diff
@@ -219,4 +221,53 @@ impl Blueprint {
...
+
+    Valid::from_iter(execution.iter(), |(op, causes)| {
+      Valid::<(), String>::from_vec_cause(causes.to_vec()).trace(op)
+    })
+    .map(|_: Vec<_>| ())
```
Comment (ologbonowiwi) [inline src/blueprint/blueprint.rs:271]: why do we need this?

File: tests/operation_spec.rs
Comment (ologbonowiwi) [inline tests/operation_spec.rs:1]: great job finding a way to test operations!

File: tests/graphql/operations/jsonplaceholder.graphql
Comment (ologbonowiwi) [inline tests/graphql/operations/jsonplaceholder.graphql:1]: can you add more test cases so we can increase coverage for this feature?

and congrats on what you've done so far, I believe it was quite tricky to test this

File: src/blueprint/blueprint.rs (hunk 19 above)
Comment (tobihans) [reply src/blueprint/blueprint.rs:271]: It converts the errors from all operations into one, and discard the expected `Vec<()>` for `()`.

File: tests/operation_spec.rs
Comment (tobihans) [reply tests/operation_spec.rs:1]: Thank you !

File: src/cli/mod.rs (hunk 18 above)
Comment (tobihans) [reply src/cli/mod.rs:7]: Solved. Thanks.

File: src/blueprint/blueprint.rs (hunk 19 above)
Comment (tobihans) [reply src/blueprint/blueprint.rs:271]: That allows to display a structured output with each failure linked to the appropriate file.

File: tests/graphql/operations/jsonplaceholder.graphql
Comment (tobihans) [reply tests/graphql/operations/jsonplaceholder.graphql:1]: Yes, I'll look at the other tests and do it. Thanks

File: src/blueprint/blueprint.rs (hunk 17 above)
Comment (tobihans) [reply src/blueprint/blueprint.rs:266]: I can use `futures::stream` for that, but that will require adding a new dependency. So I think, this is ok.

File: npm/gen-root.ts (hunk 16 above)
Comment (tobihans) [reply npm/gen-root.ts]: Yeah, this is certainly a formatting thing from my editor. THanks

File: tests/operation_spec.rs
Comment (meskill) [inline tests/operation_spec.rs:1]: this file is very similar to the graphql_spec file nearby. 

Could this two be merged together? so we can just extend our graphql tests with operations check. Anyway it's related to graphql so it's ok not to split them

File: tests/graphql/operations/jsonplaceholder.graphql [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+type Post {
+  id: Int!
+  userId: Int!
+  title: String!
+  body: String!
+  user: User @http(path: "/users/{{value.userId}}")
+}
+
+#> client-query
+query @diagnostic(count: 0) {
```
Comment (meskill) [inline tests/graphql/operations/jsonplaceholder.graphql:31]: could we borrow the approach to test errors from the graphql tests like here https://github.com/tailcallhq/tailcall/blob/0295e5e16a9c13aa93864b155c3857090e9707a5/tests/graphql/errors/test-field-already-implemented-from-Interface.graphql#L21 ?

- more implicit naming for directive and arguments
- ability to test actual error

File: src/cli/tc.rs [hunk 21]
```diff
@@ -37,18 +37,31 @@ pub fn run() -> Result<()> {
...
             display_schema(&blueprint);
           }
 
-          Ok(())
+          if let Some(out_file) = out_file_path {
+            tokio::runtime::Runtime::new()?.block_on(async { config.write_file(&out_file).await })?;
+            Fmt::display(Fmt::success(
+              &format!("Schema has been written to {}", out_file).to_string(),
+            ));
+          }
```
Comment (meskill) [inline src/cli/tc.rs:58]: this code is not related to this pr and should be removed

File: src/blueprint/blueprint.rs [hunk 22]
```diff
@@ -219,4 +221,53 @@ impl Blueprint {
...
     schema.finish().unwrap()
   }
+
+  /// Generate a dumb schema useful for validation.
+  pub fn to_validation_schema(&self) -> Schema {
+    create(self, Some(SchemaModifiers { no_resolver: true }))
+      .validation_mode(ValidationMode::Strict)
+      .disable_introspection()
+      .finish()
+      .unwrap()
```
Comment (meskill) [inline src/blueprint/blueprint.rs:231]: prefer to return Result here instead of unwrap and make a validation error from this at the caller level

File: src/blueprint/blueprint.rs [hunk 23]
```diff
@@ -219,4 +221,53 @@ impl Blueprint {
...
+          op.to_string(),
+          vec![Cause::new(format!("Cannot read file operation file {}", op))],
+        )),
+      }
+    }
+
+    Valid::from_iter(execution.iter(), |(op, causes)| {
+      Valid::<(), String>::from_vec_cause(causes.to_vec()).trace(op)
+    })
+    .map(|_| ())
```
Comment (meskill) [inline src/blueprint/blueprint.rs:271]:
```suggestion
    .unit()
```

File: src/blueprint/blueprint.rs [hunk 24]
```diff
@@ -219,4 +221,53 @@ impl Blueprint {
...
+          err
+            .locations
+            .iter()
+            .map(|loc| format!("line {} column {}", loc.line, loc.column)),
+        ))
+      })
+      .collect()
+  }
+
+  pub async fn validate_operations(&self, operations: Vec<String>) -> Valid<(), String> {
```
Comment (meskill) [inline src/blueprint/blueprint.rs:251]: we need a way to see test output for this without testing it manually. Improving operation tests for this would help

**Review by meskill — CHANGES_REQUESTED** [review]

File: src/blueprint/blueprint.rs (hunk 17 above)
Comment (meskill) [reply src/blueprint/blueprint.rs:266]: probably, using `futures_util::future::join_all` will be enough and we have that dependency.

Overall, we want to execute all the checks in parallel and not sequentially since it is expected to finish faster

File: src/blueprint/blueprint.rs [hunk 25]
```diff
@@ -219,4 +221,61 @@ impl Blueprint {
     // generation of schema cannot fail.
     schema.finish().unwrap()
   }
+
+  /// Generate a dumb schema useful for validation.
+  pub fn to_validation_schema(&self) -> Result<Schema, ValidationError<String>> {
```
Comment (meskill) [inline src/blueprint/blueprint.rs:226]: let's move all the validation stuff into separate file `blueprint/validation.rs`

File: src/blueprint/blueprint.rs [hunk 26]
```diff
@@ -219,4 +221,61 @@ impl Blueprint {
...
+      .collect()
+  }
+
+  pub async fn validate_operations(&self, operations: Vec<String>) -> Valid<(), String> {
+    match self.to_validation_schema() {
+      Ok(schema) => {
+        let mut execution = vec![];
+
+        for op in operations.iter() {
+          match tokio::fs::read_to_string(op).await {
```
Comment (meskill) [inline src/blueprint/blueprint.rs:261]: we have constraints in some environments like wasm on reading files so it's important to not have IO operations inside the core logic like blueprint. Let's move this reading out of blueprint and move it to `tc.rs` file, here just accept the operation content itself

**Review by meskill — CHANGES_REQUESTED** [review]

**Comment by ssddOnTop** [comment]
Can you resolve the conflicts

**Comment by tobihans** [comment]
> Can you resolve the conflicts

Yes, I'll do it soon. Thanks

File: src/cli/tc.rs (hunk 8 above)
Comment (tusharmath) [reply src/cli/tc.rs:152]: I think we can convert this to a log message instead. So that people have an option to enable/disable it via env variables.

File: src/blueprint/operation.rs [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+          .locations
+          .iter()
+          .map(|loc| format!("line {} column {}", loc.line, loc.column))
+          .collect(),
+      )
+    })
+    .collect()
+}
+
+pub async fn validate_operations(blueprint: &Blueprint, operations: Vec<String>) -> Valid<(), String> {
```
Comment (tusharmath) [inline src/blueprint/operation.rs:44]: Anything inside of `blueprint` shouldn't be dependent on any kind of IO.
`ValidateOperation` should take in all the operations as string, parse and validate them.

IO Operations should happen within the CLI.

File: src/blueprint/into_schema.rs [hunk 28]
```diff
@@ -177,20 +181,25 @@ fn to_type(def: &Definition) -> dynamic::Type {
...
+    schema = schema.register(to_type(def, modifiers.map(|a| a.no_resolver).unwrap_or(false)));
   }
 
   schema
 }
 
 impl From<&Blueprint> for SchemaBuilder {
   fn from(blueprint: &Blueprint) -> Self {
-    create(blueprint)
+    create(blueprint, None)
```
Comment (tusharmath) [inline src/blueprint/into_schema.rs:203]: Why are we making a change here?

File: tests/graphql/operations/jsonplaceholder.graphql [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+#> client-query
+query @diagnostic(count: 1) {
+  user(id: 1) {
+    id
+    nome
+    username
+  }
+}
```
Comment (tusharmath) [inline tests/graphql/operations/jsonplaceholder.graphql:46]: Tests in `/graphql` and `/http` exist because we have written that logical components that execute those bits. However here we are leveraging what async_graphql provides us. We should just test that when we call the methods on async_graphql, it returns results in the format we expect and that should be enough. 

The tests written in `operation.rs` should be enough for this usecase.

File: src/blueprint/into_schema.rs (hunk 28 above)
Comment (tobihans) [reply src/blueprint/into_schema.rs:203]: I'm reusing this method to create a validation schema that doesn't make actual calls. The method accepts an additional `Option<SchemaModifiers>` to modify the resulting schema. Normal behavior is then obtained by passing `None`.

File: src/cli/fmt.rs [hunk 30]
```diff
@@ -13,10 +13,6 @@ impl Fmt {
     format!("{}", meta.yellow())
   }
 
-  pub fn success(success: &String) -> String {
-    format!("{}", success.green())
-  }
```
Comment (tusharmath) [inline src/cli/fmt.rs:18]: This is a library function, why delete?

File: src/cli/tc.rs [hunk 31]
```diff
@@ -39,17 +40,31 @@ pub fn run() -> Result<()> {
       runtime.block_on(server.start())?;
       Ok(())
     }
-    Command::Check { file_paths, n_plus_one_queries, schema } => {
+    Command::Check { file_paths, n_plus_one_queries, schema, operations, out_file_path: _ } => {
```
Comment (tusharmath) [inline src/cli/tc.rs:43]:
```suggestion
    Command::Check { file_paths, n_plus_one_queries, schema, operations } => {
```

File: src/blueprint/mod.rs [hunk 32]
```diff
@@ -4,8 +4,10 @@ mod const_utils;
 mod from_config;
 mod into_schema;
 mod timeout;
+mod validation;
 
 pub use blueprint::*;
 pub use const_utils::*;
 pub use from_config::*;
 pub use timeout::GlobalTimeout;
+pub use validation::*;
```
Comment (tusharmath) [inline src/blueprint/mod.rs:13]: Rename this to `operation`. Because this will contain functions for validating operations.

File: src/cli/command.rs [hunk 33]
```diff
@@ -42,6 +42,14 @@ pub enum Command {
     /// Display schema
     #[arg(short, long)]
     schema: bool,
+
+    /// Path of the file to save final configuration in
+    #[arg(short, long("out"))]
+    out_file_path: Option<String>,
```
[the remark this answers is not in the document]
Comment (tobihans) [reply src/cli/command.rs:48]: Ok, let me remove it. It comes from me, pulling `upstream/main` to keep up to date.

File: src/cli/command.rs [hunk 34]
```diff
@@ -42,6 +42,14 @@ pub enum Command {
...
+
+    /// Operations to check
+    #[arg(short('p'), long, value_delimiter=',', num_args = 1..)]
```
Comment (tusharmath) [inline src/cli/command.rs:51]:
```suggestion
    #[arg(short, long, value_delimiter=',', num_args = 1..)]
```

**Review by tusharmath — APPROVED** [review]
